helpdesk_mgmt/tests/test_portal.py

```diff
@@ -0,0 +1,124 @@
+from odoo import http
+
+from odoo.addons.base.tests.common import HttpCaseWithUserPortal
+
+
+class TestPortal(HttpCaseWithUserPortal):
+    """Test controllers defined for portal mode.
+    This is mostly for basic coverage; we don't go as far as fully validating
+    HTML produced by our routes.
+    """
+
+    def setUp(self):
+        """Override test set-up to prepare data used in tests."""
+
+        super().setUp()
+
+        # Create a ticket submitted by our portal user.
+        self.portal_ticket = self._create_ticket(
+            self.partner_portal, "portal-ticket-title"
+        )
+
+    def test_ticket_list(self):
+        """List tickets in portal mode, ensure it contains our test ticket."""
+
+        self.authenticate("portal", "portal")
+        resp = self.url_open("/my/tickets")
+        self.assertEqual(resp.status_code, 200)
+        self.assertIn("portal-ticket-title", resp.text)
+
+    def test_ticket_form(self):
+        """Open our test ticket in portal mode."""
+
+        self.authenticate("portal", "portal")
+        resp = self.url_open(f"/my/ticket/{self.portal_ticket.id}")
+        self.assertEqual(resp.status_code, 200)
+        self.assertIn("portal-ticket-title", resp.text)
+        self.assertIn('<h4 class="page-header">History</h4>', resp.text)
+
+    def test_submit_ticket(self):
+        """Submit a ticket in portal mode."""
+
+        new_ticket_title = "portal-new-submitted-ticket-subject"
+
+        def count_new_tickets():
+            return self.env["helpdesk.ticket"].search_count(
+                [("name", "=", new_ticket_title)]
+            )
+
+        self.assertEqual(count_new_tickets(), 0)  # not found so far
+
+        self.authenticate("portal", "portal")
+        resp = self.url_open(
+            "/submitted/ticket",
+            data={
+                "category": self.env.ref("helpdesk_mgmt.helpdesk_category_1").id,
+                "csrf_token": http.WebRequest.csrf_token(self),
+                "subject": new_ticket_title,
+                "description": "portal-new-submitted-ticket-description",
+            },
+        )
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(count_new_tickets(), 1)  # new ticket found
+
+    def test_ticket_list_unauthenticated(self):
+        """Attempt to list tickets without auth, ensure we get sent back to login."""
+
+        resp = self.url_open("/my/tickets", allow_redirects=False)
+        self.assertEqual(resp.status_code, 303)
+        self.assertTrue(resp.is_redirect)
+        # http://127.0.0.1:8069/web/login?redirect=http%3A%2F%2F127.0.0.1%3A8069%2Fmy%2Ftickets
+        self.assertIn("/web/login", resp.headers["Location"])
+
+    def test_ticket_list_unauthorized(self):
+        """Attempt to list tickets without ticket ACL, ensure we get sent back
+        to /my.
+        """
+
+        self.env.ref("helpdesk_mgmt.access_helpdesk_ticket_portal").perm_read = False
+        self.authenticate("portal", "portal")
```

Comment:
```suggestion
        self.authenticate("test-basic-user", "test-basic-user")
```

Reply:
I think it is better to test it with a user without permissions in helpdesk

Reply:
In portal mode, one can't customize security groups; our "portal" demo user only has the portal security group, hence this ACL change

That test-basic-user would not have portal access at all so it is not testing the same thing ➔ I'll add but as a separate test